src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs

```diff
@@ -21,7 +22,31 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
         {
             SyntheticBoundNodeFactory factory = new SyntheticBoundNodeFactory(this, this.GetNonNullSyntaxNode(), compilationState, diagnostics);
             factory.CurrentMethod = this;
-            
+
+            // Initialize the payload root for for each kind of dynamic analysis instrumentation.
+            // A payload root is an array of arrays of per-method instrumentation payloads.
+            // For each kind of instrumentation:
+            //
+            //     payloadRoot = new T[MaximumMethodDefIndex + 1][];
+            //
+            // where T is the type of the payload at each instrumentation point, and MaximumMethodDefIndex is the 
+            // index portion of the greatest method definition token in the compilation. This guarantees that any
+            // method can use the index portion of its own method definition token as an index into the payload array.
+
+            IReadOnlyCollection<KeyValuePair<int, InstrumentationPayloadRootField>> payloadRootFields = ContainingPrivateImplementationDetailsType.GetInstrumentationPayloadRoots();
+            ArrayBuilder<BoundStatement> body = ArrayBuilder<BoundStatement>.GetInstance(2 + payloadRootFields.Count);
+            foreach (KeyValuePair<int, InstrumentationPayloadRootField> payloadRoot in payloadRootFields)
```

Comment:
When I suggested to switch to IReadonlyCollection I didn't realize that that would likely make this code non-deterministic (the order of pairs in the list is not deterministic). The code before was initializing payload roots in order of analysis kind. We can still do this without extra memory, if we switch from IReadonlyCollection  IReadOnlyDictionary<TKey, TValue>. We will be able to get the count and we can loop through possible analysis kinds, checking if there is a corresponding entry in the dictionary. This can be addressed separately.